mining/mqueue.go

```diff
@@ -0,0 +1,114 @@
+package mining
+
+import (
+	"bytes"
+	"container/heap"
+	"sort"
+
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+// MessageQueue is a priority queue of messages from different actors. Messages are ordered
+// by decreasing gas price, subject to the constraint that messages from a single actor are
+// always in increasing nonce order.
+// All messages for a queue are inserted at construction, after which messages may only
+// be popped.
+// Potential improvements include:
+// - deprioritising messages after a gap in nonce value, which can never be mined (see Ethereum)
+// - attempting to pack messages into a fixed gas limit (i.e. 0/1 knapsack subject to nonce ordering)
+type MessageQueue struct {
+	bySender queueHeap
+}
+
+// NewMessageQueue allocates and initializes a message queue.
+func NewMessageQueue(msgs []*types.SignedMessage) MessageQueue {
+	// Group messages by sender.
+	bySender := make(map[address.Address]nonceQueue)
+	for _, m := range msgs {
+		bySender[m.From] = append(bySender[m.From], m)
+	}
+
+	// Order each sender queue by nonce and initialize heap structure.
+	addrHeap := make(queueHeap, len(bySender))
+	heapIdx := 0
+	for _, nq := range bySender {
+		sort.Slice(nq, func(i, j int) bool { return nq[i].Nonce < nq[j].Nonce })
+		addrHeap[heapIdx] = nq
+		heapIdx++
+	}
+	heap.Init(&addrHeap)
+
+	return MessageQueue{addrHeap}
+}
+
+// Empty tests whether the queue is empty.
+func (mq *MessageQueue) Empty() bool {
+	return len(mq.bySender) == 0
+}
+
+// Pop removes and returns the next message from the queue, returning (nil, false) if none remain.
+func (mq *MessageQueue) Pop() (*types.SignedMessage, bool) {
+	if len(mq.bySender) == 0 {
+		return nil, false
+	}
+	// Select actor with best gas price.
+	bestQueue := &mq.bySender[0]
+
+	// Pop first message off that actor's queue
+	msg := (*bestQueue)[0]
+	if len(*bestQueue) == 1 {
+		// If the queue will become empty, remove it from the heap.
+		heap.Pop(&mq.bySender)
+	} else {
+		*bestQueue = (*bestQueue)[1:]
+		heap.Fix(&mq.bySender, 0)
```

Comment:
So this first element will be garbage collected after passing the value back?  Also I'm not sure why you don't call Pop here, since it seems like this is already ordered.  Perhaps I don't understand what Fix does (still somewhat of a go newb).

Reply:
Thanks, I've added more comments.
The actor's queue needs to be re-ordered in the heap based on the gas price of the next message it contains. `Fix` is equivalent to but faster than `x = heap.Pop(); heap.Push(x)`.